app/src/main/java/com/celzero/bravedns/service/ConnectionMonitor.kt

```diff
@@ -0,0 +1,240 @@
+        fun onNetworkDisconnected()
+        fun onNetworkConnected(networks: LinkedHashSet<Network>?)
+    }
+
+    fun removeCallBack() {
+        connectivityManager.unregisterNetworkCallback(this)
+        destroy()
+    }
+
+    private fun destroy() {
+        this.serviceHandler?.removeCallbacksAndMessages(null)
+        this.handlerThread?.quitSafely()
+        this.handlerThread = null
+        this.serviceHandler = null
+    }
+
+    override fun onAvailable(network: Network) {
+        if(DEBUG) Log.d(LOG_TAG, "ConnectionMonitor - onAvailable")
+        handleNetworkChange(isForceUpdate = false)
+    }
+
+    override fun onLost(network: Network) {
+        if(DEBUG) Log.d(LOG_TAG, "ConnectionMonitor - onLost")
+        handleNetworkChange(isForceUpdate = false)
+    }
+
+    override fun onCapabilitiesChanged(network: Network, networkCapabilities: NetworkCapabilities) {
+        if(DEBUG) Log.d(LOG_TAG, "ConnectionMonitor - onCapabilitiesChanged")
+        handleNetworkChange(isForceUpdate = false)
+    }
+
+    override fun onLinkPropertiesChanged(network: Network, linkProperties: LinkProperties) {
+        if(DEBUG) Log.d(LOG_TAG, "ConnectionMonitor - onLinkPropertiesChanged")
+        handleNetworkChange(isForceUpdate = false)
+    }
+
+    /**
+     * Called when the preference(isAddAllNetwork) is changed.
+     * if the preference is true, add only active network
+     * else, add all available network.
+     */
+    fun onUserPreferenceChanged(){
+        if (DEBUG) Log.d(LOG_TAG, "ConnectionMonitor - onUserPreferenceChanged")
+        handleNetworkChange(isForceUpdate = true)
+    }
+
+    /**
+     * Returns the network list based on the user preference.
+     * If isAddAllNetwork preference is selected then returns the list of available networks.
+     * else return the current active network.
+     */
+    fun getNetworkList(): LinkedHashSet<Network> {
+        val networks: LinkedHashSet<Network> = linkedSetOf()
+        if (persistentState.isAddAllNetworks) {
+            networks.addAll(currentNetworks)
+        } else {
+            currentActiveNetwork?.let { networks.add(it) }
+        }
+        return networks
+    }
+
+    private fun handleNetworkChange(isForceUpdate : Boolean) {
+        val message = constructMessage(if (persistentState.isAddAllNetworks) MSG_ADD_ALL_NETWORKS else MSG_ADD_ACTIVE_NETWORK, isForceUpdate)
+        serviceHandler?.removeMessages(MSG_ADD_ACTIVE_NETWORK, null)
+        serviceHandler?.removeMessages(MSG_ADD_ALL_NETWORKS, null)
+        serviceHandler?.sendMessage(message)
+    }
+
+    /**
+     * Constructs the message object for Network handler.
+     * Add the active network to the message object
+     * in case of setUnderlying network has only active networks.
+     */
+    private fun constructMessage(what: Int, isForceUpdate: Boolean): Message {
+        val message = Message.obtain()
+        message?.what = what
+        message?.obj = isForceUpdate
+        return message!!
```

Comment:
I don't get why `?` and `!!` operators are required? `Message.Obtain` will return a new Message obj always?

Reply:
Removed the null-check operators.